### DISCUSSION TITLE
Welcome to Powerwall-Dashboard Discussions!

### DISCUSSION THREAD

**Discussion by jasonacox** [jasonacox/Powerwall-Dashboard#13] [Announcements]
## 👋 Welcome Tesla Energy and Powerwall Enthusiast!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions about the Powerwall Dashboard.
  * Suggest features or graphs you would like to see added to the Dashboard.
  * Share some awesome you set up with your Dashbaord!
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by jasonacox** [comment]
I am considering adding a graph to the Dashboard to chart frequency metrics (and drift) over time (e.g. 60Hz/50Hz) as well as house voltages.  Would this be helpful?

## Going Off-Grid

I decided to test the behavior of the system in manual off-grid (Island mode) for an extended period of time.  Below is a graph showing the frequency changes that occured when going off-Grid (Island mode).   As we can see, it resulted in a drop in Powerwall and House frequency: 

<img width="1185" alt="image" src="https://user-images.githubusercontent.com/836718/158072569-cc7f6cbb-569b-4d6b-8752-51ec0edffccc.png">

Here is a graph also shows voltage differences during the same time period:

<img width="1189" alt="image" src="https://user-images.githubusercontent.com/836718/158072853-711adb72-681b-49eb-8570-be280af9428a.png">

We see the drop in frequence from nominal, 60 Hz to 59.5 Hz during Island (off grid) mode. There was also a significant separation in voltage on the two phases (118V on one leg and 125V on the other).  

## Powerwall Charged

Once the Powerwall filled up to 75%, I saw the frequency jumped from 59.5 Hz to 60 Hz.  As the charge continued, up to 95%, the frequency jump again, this time to 60.5 Hz and I noticed the solar generation started to drop.  It seems that it tries to hold the Powerwall battery level at 98%.  It seems to alternate between draining the Powerwall battery until it drops below 98% and then ramps up solar generation to charge it back up to 98%: 

<img width="1185" alt="image" src="https://user-images.githubusercontent.com/836718/158078997-1e3b0bb7-91e7-4771-8f02-b476280ee29a.png">

<img width="1053" alt="image" src="https://user-images.githubusercontent.com/836718/158079075-86a396a2-3fb8-4635-b77a-133d61d0aba6.png">

**Comment by cwagz** [comment]
Very cool! I need to try this out.

**Comment by jasonacox** [comment]
I added this to the default [dashboard.json](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/dashboard.json) setup:

<img width="1099" alt="image" src="https://user-images.githubusercontent.com/836718/161393960-87d6c8f1-2f00-4a5b-b201-3ced1fbb44bc.png">


This requires InfluxDB updates to downsample the frequency and voltage data for long term display:


```bash
  # make sure you have a good backup first

  # to get new influxdb.sql dashboard.json and telegraf.conf files
  git pull 

  # add downsample 
  docker exec -it influxdb influx -import -path=/var/lib/influxdb/influxdb.sql

  # update telegraf to pull freq and voltage data
  docker-compose -f powerwall.yml up -d
```

**Comment by youzer-name** [comment]
I have a pair of Powerwall 2's, but not the + version, and when I tried to implement the latest changes I didn't get any new data. Since I don't have the strings data I commented that line out of the telegraf.conf and ignored the related database queries for strings. 

For the frequency and voltage data, I see "http://pypowerwall:8675/freq" in the Telegraf conf. When I try to run that in a browser, I get a return of "ERROR!". I see in the docs however that there is a URL for http://pypowerwall:8675/vitals. That does return some data in what looks like JSON format.

I tried using that URL in the telegraf.conf but it made a mess of my data, creating columns (or are they field keys? I'm not proficient with InfluxDB yet) for each of the nodes of the JSON, but they don't seem to be the right fields based on what I see in the dashboard.json. I'm going to try to drop the retention policy and see if that cleans up the database. 

So is the /freq URL and the extra data for Powerwall+ only?

**Reply by jasonacox** [reply]
Make sure you have the latest pypowerwall container installed.  If you can't get the `/freq` data that indicates you may have an older version. 

Browse to this: http://pypowerwall:8675/stats and it should look something like this:

```json
{"pypowerwall": "0.4.0 Proxy t5", "gets": 122723, "errors": 50, "timeout": 0, "uri": {"/csv": 15904, "/aggregates": 21208, "/soe": 21208, "/freq": 21207, "/temps/pw": 21208, "/strings": 21560, "/stats": 30, "/vitals": 398}, "ts": 1647654172, "start": 1647548133, "clear": 1647548133}
```
If you see a lower version of pypowerwall (0.4.0 Proxy t5), install an upgrade:

```bash
docker pull jasonacox/pypowerwall
```

I would not telegraf the /vitals API into InfluxDB.  It is a big mess. :)

**Reply by cwagz** [reply]
> ```shell
> docker pull jasonacox/pypowerwall
> ```

I ran this and restarted the Docker instance, but I am still seeing version 0.3.0. Any ideas?

Thanks.

**Reply by jasonacox** [reply]
Ok, you might need to delete the image and force the pull.  Here is a script that will delete and rebuild pypowerwall (run from directory where you have powerwall.yml):

```bash
#!/bin/bash

# make sure you have a good backup first

echo "Upgrade pyPowerwall to Latest"

echo "Stopping..."
docker stop pypowerwall

echo "Deleting..."
docker rm pypowerwall
docker images | grep pypowerwall | awk '{print $3}' | xargs docker rmi -f

echo "Rebuilding..."
docker-compose -f powerwall.yml up -d

echo "Done"
```

**Reply by cwagz** [reply]
That worked. Thanks!

**Comment by youzer-name** [comment]
That was it. I'm not sure why it didn't update when I did a git pull, but I just recreated it with the 'pull latest image' option using Portainer and it is now on 0.4.0 t5. Thanks!

My setup is a bit customized so I have to either pull the latest changes into my config files manually, or edit the newly released files each time to get it set up right. My containers and hosts have different names and use some alternate ports because I already have other InfluxDB and Telegraf hosts running, and I'm using an existing Grafana container for the graphs.

It looks like the extra columns (tags? keys? Influx confuses me) from pulling /vitals into the DB are gone from the latest shard, so I guess no harm done.

**Comment by jasonacox** [comment]
That's great news @youzer-name !  Thanks for the confirmation and insight.  Yes, unless you created downsampling `CONTINUOUS QUERY` jobs for those raw /vitals data points, they will expire out of InfluxDB in 3 days (via the initial [influxdb.sql](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/influxdb/influxdb.sql) setup).  

I have been thinking about the best way to offer up an easier upgrade path for anyone wanting to get the latest dashboard and features.  I haven't considered the custom port config that you mention but that may be a popular setup that could be accomodated.  If we update the [setup.sh](https://github.com/jasonacox/Powerwall-Dashboard/blob/main/setup.sh) to prompt for those ports and write it to a state file, it could be reused in an 'upgrade.sh' script to pull latest.  Would that be a worthwhile investment or do you think most people would manually upgrade anyway?

**Comment by youzer-name** [comment]
That would definitely be useful to me, and I'm probably not the only one running multiple databases or telegraf instances. I'd need the hostname, container name, and ports to be configurable either in the setup.sh or using a .env file where the values could be set.

**Comment by mikkyo9** [comment]
I got my docker instance all updated with the latest release, and the new Dashboard with freq support.
I have 3 Powerwall2s, not the Plus, and still can see the freqs, and voltages, which is nice.
I see 3 interverter sections TEPINV--1081100-49-J--XXXXXXXXX, which seem to correspond to the 3 strings going into the inverter.
Does the PINV_Vout from each inverter section show the string voltage or is that something else?
"PINV_VSplit1": 120.2,
"PINV_VSplit2": 121.5,
"PINV_Vout": 241.5,

**Reply by jasonacox** [reply]
Nice!  Do you have three different TEPINV sections?  That may mean that it maps each of your strings to a different inverter.

PINV_V* key is the voltage output from the inverters.  The VSplit number are just the breakout of the two phases (L1 and L2) with Vout being the combined.

**Reply by mikkyo9** [reply]
Yes, 3 TEPINV, and 3 TEPOD sections.

**Reply by mikkyo9** [reply]
Looks like 1 inv per PW, I think it must be in the PW.
I see the TEPOD also has the POD_ActiveHeating status, which might be good to show on the graphs as well.

**Reply by jasonacox** [reply]
That's a great idea!  I show the temp from TETHC (parent of TEPOD) but it would be good to show if active heating is turned on.

**Comment by jasonacox** [comment]
Current Solar Generation - 27 Mar 2022:

<img width="963" alt="image" src="https://user-images.githubusercontent.com/836718/160303148-b8f563a0-1ec9-4fde-8bc4-af0358b200ad.png">

<img width="611" alt="image" src="https://user-images.githubusercontent.com/836718/160303229-a417c066-c507-4fbf-852b-0f3531e5e4a6.png">

<img width="1113" alt="image" src="https://user-images.githubusercontent.com/836718/161393841-1349a93c-8876-4829-abc4-546bfe492d61.png">
